pkg/kademlia/workers.go

```diff
@@ -0,0 +1,164 @@
+	mu             *sync.Mutex
+	maxResponse    time.Duration
+	cancel         context.CancelFunc
+	nodeClient     node.Client
+	find           dht.NodeID
+	workInProgress int
+	k              int
+}
+
+func newWorker(ctx context.Context, rt *RoutingTable, nodes []*proto.Node, nc node.Client, target dht.NodeID, k int) *worker {
+	t := new(big.Int).SetBytes(target.Bytes())
+
+	pq := func(nodes []*proto.Node) PriorityQueue {
+		pq := make(PriorityQueue, len(nodes))
+		for i, node := range nodes {
+			bnode := new(big.Int).SetBytes([]byte(node.GetId()))
+			pq[i] = &Item{
+				value:    node,
+				priority: new(big.Int).Xor(t, bnode),
+				index:    i,
+			}
+
+		}
+		heap.Init(&pq)
+
+		return pq
+	}(nodes)
+
+	return &worker{
+		pq:             pq,
+		mu:             &sync.Mutex{},
+		maxResponse:    0 * time.Millisecond,
+		nodeClient:     nc,
+		find:           target,
+		workInProgress: 0,
+		k:              k,
+	}
+}
+
+func (w *worker) work(ctx context.Context, ch chan []*proto.Node) error {
+	// grab uncontacted node from working set
+	// change status to inprogress
+	// ask node for target
+	// if node has target cancel ctx and send node
+	for {
+		if ctx.Err() != nil {
+			return nil
+		}
+		n := w.getWork()
+		if n == nil {
+			continue
+		}
+
+		nodes := w.lookup(ctx, n)
+		if nodes == nil {
+			continue
+		}
+
+		ch <- nodes
+
+		if err := w.update(nodes); err != nil {
+			return err
+		}
+		continue
+	}
+
+}
+
+func (w *worker) getWork() *proto.Node {
+	w.mu.Lock()
+	if w.pq.Len() <= 0 {
+		w.mu.Unlock()
+		timeout := defaultTimeout
+		if timeout < (2 * w.maxResponse) {
+			timeout = 2 * w.maxResponse
+		}
+
+		time.AfterFunc(timeout, w.cancel)
+		return nil
+	}
+	defer w.mu.Unlock()
+	if w.pq.Len() <= 0 {
+		return nil
+	}
+
+	w.workInProgress++
+	return w.pq.Pop().(*Item).value
+}
+
+func (w *worker) lookup(ctx context.Context, node *proto.Node) []*proto.Node {
+	start := time.Now()
+	if node.GetAddress() == nil {
+		return nil
+	}
+
+	nodes, err := w.nodeClient.Lookup(ctx, *node, proto.Node{Id: w.find.String()})
+	if err != nil {
+		// TODO(coyle): I think we might want to do another look up on this node or update something
+		// but for now let's just log and ignore.
+		log.Printf("Error occured during lookup for %s on %s :: error = %s", w.find.String(), node.GetId(), err.Error())
+		return []*proto.Node{}
+	}
+
+	latency := time.Now().Sub(start)
+	if latency > w.maxResponse {
+		w.maxResponse = latency
+	}
+
+	return nodes
+}
+
+func (w *worker) update(nodes []*proto.Node) error {
+	if len(nodes) == 0 {
+		return WorkerError.New("nodes must not be empty")
+	}
+	t := new(big.Int).SetBytes(w.find.Bytes())
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	for _, v := range nodes {
+		w.pq.Push(&Item{
+			value:    v,
+			priority: new(big.Int).Xor(t, new(big.Int).SetBytes(w.find.Bytes())),
+		})
+	}
+	// only keep the k closest nodes
+	if len(w.pq) > w.k {
+		w.pq = w.pq[:w.k]
+	}
+	// reinitialize heap
+	heap.Init(&w.pq)
```

Comment:
so, the heap invariant is actually kept by the w.pq[:w.k] slicing believe it or not.

from https://en.wikibooks.org/wiki/Data_Structures/Min_and_Max_Heaps, the only thing that is true of w.pq is that:

```
array[n] <= array[2*n]   // parent element <= left child
```
and
```
array[n] <= array[2*n+1] // parent element <= right child
```

these two facts mean that slicing stuff off will maintain the heap invariant, but because it doesn't say that `array[2*n] <= array[2*n+1]`, slicing stuff off may remove things that are within the closest k.